### DISCUSSION TITLE
Is there a way we can see/log the actual forward proxy request's details?

### DISCUSSION THREAD

**Discussion by ChintanRaval** [dotnet/yarp#2205] [Q&A] [answered]
```sh
dotnet 7
yarp 2.0.1
```

Hello!

Just like the `app.UseHttpLogging()` logs the request & response and you can control how much to log and see there, is there a way I can also log the proxy request that is being forwarded by YARP? Maybe it's [this object](https://github.com/microsoft/reverse-proxy/blob/1c2ee13fae896ad96e6baf5f58fb0510747d5ed3/src/ReverseProxy/Transforms/RequestTransformContext.cs#L26) `RequestTransformContext.ProxyRequest`? Or maybe it's some other object with the details? 

However, i am struggling to log the details of the actual request that is being sent out by YARP. The closest I can get is by referencing the `ReverseProxy.Metrics.Sample`'s `HttpClientTelemetryConsumer`'s `OnRequestStart` [method](https://github.com/microsoft/reverse-proxy/blob/1c2ee13fae896ad96e6baf5f58fb0510747d5ed3/samples/ReverseProxy.Metrics.Sample/HttpClientTelemetryConsumer.cs#L12-L13) which har some of the details coming in as arguments which I could log by wiring up that class in the middleware (`builder.Services.AddTelemetryConsumer<HttpClientTelemetryConsumer>();`

It seems like a simple thing to see when using YARP, but I think I am definitely overlooking something that may be obviously available - so some help here is really appreciated 🙂 

P.S.: Yes - I have tried searching the YARP repo, this discussion forum, YARP's documentation, search engine & chatGPT 😅

**Comment by ChintanRaval** [comment]
UPDATE: I think I may have found _one_ way to access the forward proxy request by using a Custom Transformer as described [here](https://github.com/microsoft/reverse-proxy/blob/1c2ee13fae896ad96e6baf5f58fb0510747d5ed3/samples/ReverseProxy.Direct.Sample/Startup.cs#L108) and then use `HttpTransformer.TransformRequestAsync(...)` method to access the `HttpRequestMessage proxyRequest` argument. 

However, this seems to require moving away from the regular `app.MapReverseProxy()` call, is there a way one can log/see/access the forward request object without using direct forwarding?

**Comment by MihaZupan** [comment, marked as answer]
What information about the request are you looking to log?
Things like Uri/Host/Headers, or also the request body of POST requests?

There is nothing similar to `UseHttpLogging` built in, but most of the information on the outgoing request should be very similar to the original request that came in (minus Uri and header modifications done with transforms).

> However, this seems to require moving away from the regular app.MapReverseProxy() call, is there a way one can log/see/access the forward request object without using direct forwarding?

You don't have to write a whole `HttpTransformer` or move to direct forwarding to do so.
E.g.
```c#
services.AddReverseProxy()
    .AddTransforms(transformBuilderContext =>
    {
        transformBuilderContext.AddRequestTransform(context =>
        {
            _logger.LogInformation("...", context.ProxyRequest.RequestUri.AbsoluteUri);

            return ValueTask.CompletedTask;
        });
    });
```
See https://microsoft.github.io/reverse-proxy/articles/transforms.html


If you are interested in logging the request body, I think the best option is to let [ASP.NET Core's Http Logging](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/http-logging) do it for you.

**Reply by ChintanRaval** [reply]
thanks @MihaZupan that helps and solved my problem 🙂

**Comment by jakebanks** [comment]
Is there any way to see the egress request **after** transforms? I am working with @ChintanRaval and we are now looking to compare headers on the ingress request vs headers on the egress request.

One thing I tried was using a middleware but the request on the `HttpContext` is the ingress request.

Thank you.

**Reply by MihaZupan** [reply]
You could try inserting that logic as the last request transform, or write a `DelegatingHandler` that by definition runs after all request transforms.

**Reply by jakebanks** [reply]
Thanks @MihaZupan. We actually realized that comparing header values ingress/egress is essentially 'testing YARP' and we trust the headers are being copied over, so I won't be pursuing this line of thought further.
